add staging deployment

DeusAvalon opened the pull request:
This PR adds staging CD to deploy application to https://main.switchboard.k8s.sidestream.tech

DeusAvalon committed:
add staging deployment

DeusAvalon committed:
fix workflow

DeusAvalon committed:
rm chamber command

DeusAvalon committed:
add jwt_secret

DeusAvalon committed:
disable healthcheck

DeusAvalon committed:
fix Dockerfile with prisma db push

DeusAvalon committed:
fix prisma command

DeusAvalon committed:
add build arg to Dockerfile

DeusAvalon commented:
Application is reachable under https://main.switchboard.k8s.sidestream.tech and https://main.switchboard.k8s.sidestream.tech/api/

When running the test query `{ coreUnits { id } }` i get following response:

```
{
  "errors": [
    {
      "message": "\nInvalid `prisma.coreUnit.findMany()` invocation:\n\n\nerror: Error validating datasource `db`: the URL must start with the protocol `postgresql://` or `postgres://`.\n  -->  schema.prisma:8\n   | \n 7 |   provider = \"postgresql\"\n 8 |   url      = env(\"DATABASE_URL\")\n   | \n\nValidation Error Count: 1",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "coreUnits"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "coreUnits": null
  }
}
```

It looks like `DATABASE_URL` is not correctly being parsed into the `prisma-schema`? But `npx prisma db push` seems to run smoothly as the ENV var is correctly exposed in the container see:

![image](https://user-images.githubusercontent.com/42768942/228879915-49607fde-97b2-41e5-8ace-af805b0fd399.png)

any idea @valiafetisov ?

valiafetisov commented:
Do you maybe also need to make prisma output location configurable `output = env("PRISMA_CLIENT_OUTPUT")`, [as here](https://github.com/sidestream-tech/t4f-gxmanager/blob/0ee1b205946a307a562f5709b6e4bc82a11ad094/prisma/schema.prisma#L7-L8)? I see that you actually don't need to change prisma `provider` to postgres  (which is a bit surprising nevertheless)

valiafetisov commented on frontend/Dockerfile:
```diff
@@ -30,7 +30,8 @@ ENV NUXT_HOST=0.0.0.0
 
-ENV DATABASE_URL=file:./db.sqlite
+ARG DATABASE_URL
+ENV DATABASE_URL=${DATABASE_URL}
```
Frontend actually doesn't have a database 🙂

valiafetisov commented on docker-compose.yaml:
```diff
@@ -22,6 +24,8 @@ services:
     expose:
       - 3000
+    environment:
+      DATABASE_URL: file:./db.sqlite
```
Same here

DeusAvalon committed:
add prisma client output env

DeusAvalon commented:
Adding the `PRISMA_CLIENT_OUTPUT` to the prisma.schema as mentioned in your comment and providing it via Env var results in `npx prisma db push` failing now see:

```
Prisma schema loaded from prisma/schema.prisma
Datasource "db": PostgreSQL database "postgres", schema "public" at "postgresql:5432"

The database is already in sync with the Prisma schema.

Running generate... (Use --skip-generate to skip the generators)
Running generate... - Prisma Client
✔ Generated Prisma Client (4.11.0 | library) to ./.output/server/node_modules/.p
risma/client in 168ms


> powerhouse-switchboard@0.0.1 start
> NODE_ENV=production vite-node src/index.ts

/app/node_modules/.prisma/client/index.js:3
    throw new Error(
          ^
Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    at new PrismaClient (/app/node_modules/.prisma/client/index.js:3:11)
    at /app/src/database.ts:36:20
    at ViteNodeRunner.directRequest (file:///app/node_modules/vite-node/dist/client.mjs:315:5)
    at ViteNodeRunner.cachedRequest (file:///app/node_modules/vite-node/dist/client.mjs:159:14)
    at ViteNodeRunner.dependencyRequest (file:///app/node_modules/vite-node/dist/client.mjs:207:14)
    at /app/src/context.ts:7:31
    at ViteNodeRunner.directRequest (file:///app/node_modules/vite-node/dist/client.mjs:315:5)
    at ViteNodeRunner.cachedRequest (file:///app/node_modules/vite-node/dist/client.mjs:159:14)
    at ViteNodeRunner.dependencyRequest (file:///app/node_modules/vite-node/dist/client.mjs:207:14)
    at /app/src/server.ts:15:31
```

valiafetisov commented on docker-compose.yaml:
```diff
@@ -10,6 +10,9 @@ services:
       - ./api/.env
+    environment:
+      DATABASE_URL: file:./db.sqlite
+      PRISMA_CLIENT_OUTPUT: "/app/.output/server/node_modules/.prisma/client"
```
This is path is valid for nuxt, but this service (`api`) doesn't use nuxt or build anything into `.output` folder, so the path is ultimately incorrect.  Or does prisma recursively creates all those folders that doesn't exist and doesn't care about them? Maybe specifying it didn't really helped here, but instead created another problem

DeusAvalon replied:
Yeah that seems to be the case so i removed it for now. Unfortunatly after a lot of try and error im still not sure what values needs to be added.

DeusAvalon committed:
Merge branch 'main' into add-staging-deployment

DeusAvalon committed:
rm file output car

DeusAvalon committed:
add db-string directly

DeusAvalon commented:
Even after harcoding the `DATABASE_URL` into the `prisma.schema` the graphql playground still throws following error:
```
{
  "errors": [
    {
      "message": "\nInvalid `prisma.coreUnit.findMany()` invocation:\n\n\nerror: Error validating datasource `db`: the URL must start with the protocol `postgresql://` or `postgres://`.\n  -->  schema.prisma:8\n   | \n 7 |   provider = \"postgresql\"\n 8 |   url      = \"postgres://postgres:postgres@postgresql:5432/postgres?sslmode=disable\"\n   | \n\nValidation Error Count: 1",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "coreUnits"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "coreUnits": null
  }
}
```

The `DATABASE_URL` is 100% working tho as mentioned before, `npx prisma db push` works without any issues during the startup of the api container.

DeusAvalon committed:
add healthcheck back

DeusAvalon committed:
fix prisma schema

valiafetisov committed:
fix URL must start with the protocol

valiafetisov commented:
Was able to reproduce error locally via docker-compose. There were 3 problems:
- `postinstall` script inside `api/package.json` – this automatically executed `prisma generate` during image build (which obviously can not build properly as env variable is missing as well as the database)

- `provider` inside `api/prisma/schema.prisma` was set to `sqlite` – in this version of prisma it seems it's no longer possible to use postgres url with this provider

- `DATABASE_URL` env variable contained string *with quotes* – prisma wasn't able to unquote and interpret it leading to:
    ```
    Error validating datasource `db`: the URL must start with the protocol `postgresql://` or `postgres://`.
    ```
All of them are solved now via 4193a77. Please apply the same fix to staging!

DeusAvalon commented:
> provider inside api/prisma/schema.prisma was set to sqlite – in this version of prisma it seems it's no longer possible to use postgres url with this provider

Yep thats correct thats why im replacing the provider `sqlite` with `postgres` in the deployment pipeline, so that the devs are still able to use sqlite for local development without needing to touch the `prisma.schema` 

```
    - run: sed --in-place 's/sqlite/postgresql/g' api/prisma/schema.prisma
```

DeusAvalon commented on .github/workflows/build-and-deploy-staging.yaml:
```diff
@@ -0,0 +1,83 @@
+      id: vars
+      run: echo "git_hash_short=$(git rev-parse --short ${GITHUB_SHA})" >> $GITHUB_OUTPUT
+
+    - run: sed --in-place 's/sqlite/postgresql/g' api/prisma/schema.prisma
```
@valiafetisov see here :)

DeusAvalon committed:
switch provider back and trigger deployment

DeusAvalon committed:
rm quotes arround connection string

DeusAvalon commented:
Deploying the your latest changes still results in
```
[14:18:38.012] ERROR (153): [DATABASE] Prisma error
    module: "src/database.ts"
    message: "\nInvalid `prisma.user.findFirst()` invocation:\n\n\nerror: Error validating datasource `db`: the URL must start with the protocol `postgresql://` or `postgres://`.\n  -->  schema.prisma:8\n   | \n 7 |   provider = \"postgresql\"\n 8 |   url      = env(\"DATABASE_URL\")\n   | \n\nValidation Error Count: 1"
    target: "user.findFirst"
```

I made sure that the connection string is not in quotes as well.
```
root@switchboard-api-686d8f5bc7-bqd2l:/app# env | grep 'DATABASE'
DATABASE_URL=postgres://postgres:postgres@postgresql:5432/postgres?sslmode=disable
```
Any ideas? Its weird that the same approach with the `prisma.schema` works for t4f and hOS @valiafetisov :/

valiafetisov committed:
move sed command into docker

valiafetisov committed:
remove hardcoded database url

valiafetisov committed:
fix test

valiafetisov committed:
fix playground iframe url

valiafetisov commented:
- I've moved `sed` command into dockerfile via 96fdf0a (so the trick is reproducible locally via docker-compose)
- Removed hardcoded `file:./db.sqlite` from the codebase via 602ec93 (which did the trick and removed the error)
- And fixed frontend iframe address from `/api` to `/api/`

The staging is deployed and running on https://main.switchboard.k8s.sidestream.tech

@DeusAvalon 
- I guess you need to change triggers and clean up?
- Could you please check if you've got enough rights to set up the secrets on the main repo?
- Also, is it possible to reopen this PR against main repo while keeping the CD running or that's not how it works? 
- And lastly, could you please enable sign up, like in the latest docker-compose file? To test, you can try to execute
```gql
mutation {
  signUp(user: { username: "test", password: "test" }) {
    token
  }
}
```
In the [playground](https://main.switchboard.k8s.sidestream.tech/graphql-playground)

DeusAvalon committed:
moved sed command back to cd

DeusAvalon committed:
enable auth signing and cleanup

DeusAvalon committed:
revert triggers

DeusAvalon commented:
> I guess you need to change triggers and clean up?

Yep, i moved the `sed` command back into the CD, cleaned up env vars and reverted the deployment trigger!

> Could you please check if you've got enough rights to set up the secrets on the main repo?

Yes, i do look everyday into the maker repo to check if i have admin access. So far no luck tho :(

> Also, is it possible to reopen this PR against main repo while keeping the CD running or that's not how it works?

So that we have the CI/CD Pipeline running for both repos? This would def be possible but would require that this fork is always up-to-date with the source repo before deploying.

> And lastly, could you please enable sign up, like in the latest docker-compose file? To test, you can try to execute

Done, and works as expected i guess. Enabled signup, ran the query and received a token.

valiafetisov commented:
> i moved the sed command back into the CD

Why do you need to do that? The command is already executed during build, so it doesn't need to be added to the CI, see here:
https://github.com/sidestream-tech/switchboard-boilerplate/blob/6bd99a46f3c3790f61e63180b6080da3f0769873/api/Dockerfile#L21

The proposal is then remove the same command from the CI, but leave it in the Dockerfile

> So that we have the CI/CD Pipeline running for both repos? This would def be possible but would require that this fork is always up-to-date with the source repo before deploying.

No, I meant that a PR opened against [makerdao-ses/switchboard-boilerplate](https://github.com/makerdao-ses/switchboard-boilerplate) will still be branched from [sidestream-tech/switchboard-boilerplate](https://github.com/sidestream-tech/switchboard-boilerplate). But I guess for the CI it doesn't matter where it's branched from, but rather where the PR is opened. You can ignore this suggestion. Let's just wait for the Wouter to give you the rights to set secrets.

DeusAvalon commented:
> The proposal is then remove the same command from the CI, but leave it in the Dockerfile

But that would mean the Developers can not build based on `sqlite` during the development using the Dockerfile or without making changes to it? Or are the devs not using `sqlite` at all and only use the provided `docker-compose.yml`? If this is the case there isnt even a need to replace the provider and instead just hardcode `postgresql` in it, no? 👀 


> But I guess for the CI it doesn't matter where it's branched from, but rather where the PR is opened. You can ignore this suggestion. Let's just wait for the Wouter to give you the rights to set secrets.

Yeah i hope it will happen soon that i will get access :)

valiafetisov commented:
> But that would mean the Developers can not build based on sqlite during the development using the Dockerfile

The whole point of using sqlite is to not touch docker. But if docker is required, `docker-compose` file should be in my opinion as close to production as possible, so that the developer/PM can test production/staging before merging the branch. In any case `Dockerfile` builds app into production state (at least with no `target` specified) so it's not directly useable for development, but rather for functional testing, which is why I moved it to the `Dockerfile`

DeusAvalon committed:
move sed command back to Dockerfile

DeusAvalon commented:
> The whole point of using sqlite is to not touch docker

Yes we are on the same page in this regard! I didnt know that the `docker-compose.yml` was meant as a "final" check for the Developers before merging, as the workflow in `t4f` and `hOS` is not relying on this extra step and therefor the confusion from my side! :) I reverted the changes back

valiafetisov reviewed (approved):
Code approved! Is it possible to merge this PR into our fork, but then still receive commits from the upstream? Will they be automatically deployed?

DeusAvalon commented:
I pretty much assume yes as the commits from the upstream will be rebased, which the CI consideres as `push` to main and should trigger an automatic deployment.
